### PR TITLE
DDP Client initial refactoring and cleanup

### PR DESCRIPTION
**Pull request by stubailo**
This PR is essentially cleaning up and modernizing the `ddp-client` package in all the obvious ways:

1. Remove `LivedataTest` and use modules to import into testing code
2. Remove `underscore` everywhere in the package except `livedata_connection.js`
3. Use classes, class methods, and inheritance instead of functions and `_.extend`
4. Remove self=this, and convert to arrow functions for callbacks
5. Factor out `MethodInvoker` into a separate file
6. Move all small junk into `namespace.js`
7. Removed an extra function that was generating method IDs in favor of a much simpler approach

This should make the DDP Client codebase much easier to work with for future improvements!

I think the next big improvement would be removing `underscore` from the client bundle of a bare Meteor app.

### Todo

- [x] ~Switch to parameter destructuring for default options~
- [x] Ensure `Object.entries` is polyfilled (switched to `Object.keys` instead – @benjamn)
- [x] Switch to more concise require statement
- [x] Use `String#startsWith`
- [x] Better comments around weird import stuff
- [x] Use `Object.assign` when mutating
- [x] Use `Array.from` instead of `Array.prototype.slice.call`
- [x] Check up on `this` inside arrow functions, for example here: https://github.com/meteor/meteor/pull/9338/files/8fb63d1b868f639bc06b4164cf0cac50fee45eac..45928b2cb2219c73d33eedf091ab8493aa8a6946#diff-9c102a467818ed87f1893a47f91bbf76R180

### CODE REVIEW

File: packages/ddp-client/common/urlHelpers.js [hunk 1]
```diff
@@ -1,18 +1,19 @@
-import { LivedataTest } from './namespace';
+import { LivedataTest } from './namespace.js';
 import { Random } from 'meteor/random';
 
 // XXX from Underscore.String (http://epeli.github.com/underscore.string/)
-var startsWith = function(str, starts) {
+function startsWith(str, starts) {
```
Comment (benjamn) [inline packages/ddp-client/common/urlHelpers.js]: I think we can just use `string.startsWith(prefix)` and string.endsWith(suffix)`, which is a `String` method that we polyfill.

**Comment by benjamn** [comment]
@stubailo Some of my comments are rightfully "outdated" (like the requests for additional comments), but others are worth considering (just click on the expando to see them).

File: packages/ddp-client/common/stream_client_common.js [hunk 2]
```diff
@@ -129,46 +121,41 @@ export default class StreamClientCommon {
...
-      self.options.retry ||
+      this.options.retry ||
       (maybeError && maybeError.errorType === 'DDP.ForcedReconnectError')
     ) {
-      timeout = self._retry.retryLater(
-        self.currentStatus.retryCount,
-        _.bind(self._retryNow, self)
+      timeout = this._retry.retryLater(
+        this.currentStatus.retryCount,
+        this._retryNow.bind(this)
```
[the remark this answers is not in the document]
Comment (abernix) [reply packages/ddp-client/common/stream_client_common.js]: Actually, I think I now see how this code works.  I was reviewing commit by commit.  Likely false alarm.

Sidenote: Thanks for the excellent tactics in breaking out of commits.  It did make the totality of this quite reasonable! :smile:

File: packages/ddp-client/server/stream_client_nodejs.js [hunk 3]
```diff
@@ -164,53 +153,53 @@ class ClientStream extends StreamClientCommon {
...
       }, 'stream connect callback')
     );
 
-    var clientOnIfCurrent = function(event, description, f) {
-      self.client.on(
+    var clientOnIfCurrent = (event, description, f) => {
+      this.client.on(
         event,
-        Meteor.bindEnvironment(function() {
+        Meteor.bindEnvironment(() => {
```
[the remark this answers is not in the document]
Comment (stubailo) [reply packages/ddp-client/server/stream_client_nodejs.js]: Turns out we didn't use the `this` at all, so I replaced it with something else

File: packages/ddp-client/common/livedata_connection.js [hunk 4]
```diff
@@ -41,381 +41,382 @@ class MongoIDMap extends IdMap {
...
 // or the server migrating us).
-var Connection = function(url, options) {
-  var self = this;
-  options = _.extend(
-    {
-      onConnected: function() {},
-      onDDPVersionNegotiationFailure: function(description) {
-        Meteor._debug(description);
+class Connection {
+  constructor(url, options) {
```
[the remark this answers is not in the document]
Comment (stubailo) [reply packages/ddp-client/common/livedata_connection.js]: @benjamn I'm not sure that would be easier to read - can you provide a code sample? (to be honest, I'm not sure how to accomplish this case, for example, with default values and destructuring)

**Comment by stubailo** [comment]
Moving comment here: 

@benjamn I'm not sure that would be easier to read - can you provide a code sample? (to be honest, I'm not sure how to accomplish this case, for example, with default values and destructuring)

**Comment by stubailo** [comment]
Thanks for finishing this up and merging, Ben! Hope the package is easier to work with after these changes :]

**Pull request merged**